syntax = "proto3";

package bookstore.order;

import "common.proto";

option go_package = "github.com/bookstore/contracts/gen/go/order;order";
option java_package = "com.bookstore.order";
option java_multiple_files = true;

service OrderService {
    rpc CreateOrder (CreateOrderRequest) returns (CreateOrderResponse);
    rpc GetOrder (GetOrderRequest) returns (GetOrderResponse);
    rpc UpdateOrderStatus (UpdateOrderStatusRequest) returns (UpdateOrderStatusResponse);
    rpc ListOrders (ListOrdersRequest) returns (ListOrdersResponse);
    rpc GetOrdersByUser (GetOrdersByUserRequest) returns (GetOrdersByUserResponse);
    
    // Health check
    rpc Health(bookstore.common.Empty) returns (bookstore.common.HealthStatus);
}

message CreateOrderRequest {
    string user_id = 1;
    repeated OrderItem items = 2;
    string payment_method = 3;
    string address = 4;
}

message OrderItem {
    string product_id = 1;
    int32 quantity = 2;
    bookstore.common.Money unit_price = 3;
}

message CreateOrderResponse {
    string order_id = 1;
    string status = 2; // CREATED, CONFIRMED, CANCELLED
    bookstore.common.Money total_amount = 3;
}

message GetOrderRequest {
    string order_id = 1;
}

message GetOrderResponse {
    string order_id = 1;
    string user_id = 2;
    repeated OrderItem items = 3;
    string payment_method = 4;
    string address = 5;
    string status = 6; // CREATED, CONFIRMED, CANCELLED
    bookstore.common.Money total_amount = 7;
    int64 created_at = 8;  // Unix timestamp
}

message UpdateOrderStatusRequest {
    string order_id = 1;
    string status = 2;
}

message UpdateOrderStatusResponse {
    string order_id = 1;
    string status = 2;
    bool success = 3;
    string message = 4;
}

message ListOrdersRequest {
    int32 page = 1;
    int32 page_size = 2;
}

message ListOrdersResponse {
    repeated OrderSummary orders = 1;
    int32 total_count = 2;
    int32 page = 3;
    int32 page_size = 4;
}

message GetOrdersByUserRequest {
    string user_id = 1;
    int32 page = 2;
    int32 page_size = 3;
}

message GetOrdersByUserResponse {
    repeated OrderSummary orders = 1;
    int32 total_count = 2;
    string user_id = 3;
}

message OrderSummary {
    string order_id = 1;
    string user_id = 2;
    bookstore.common.Money total_amount = 3;
    string status = 4;
    int64 created_at = 5;  // Unix timestamp
    repeated OrderItem items = 6;  // Opcional: items de la orden
}